sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonParsingOptionsSuite.scala

```diff
@@ -93,23 +93,45 @@ class JsonParsingOptionsSuite extends QueryTest with SharedSQLContext {
     assert(df.first().getLong(0) == 18)
   }
 
-  // The following two tests are not really working - need to look into Jackson's
-  // JsonParser.Feature.ALLOW_NON_NUMERIC_NUMBERS.
-  ignore("allowNonNumericNumbers off") {
-    val str = """{"age": NaN}"""
-    val rdd = spark.sparkContext.parallelize(Seq(str))
-    val df = spark.read.json(rdd)
-
-    assert(df.schema.head.name == "_corrupt_record")
+  test("allowNonNumericNumbers off") {
+    // non-quoted non-numeric numbers don't work if allowNonNumericNumbers is off.
+    var testCases: Seq[String] = Seq("""{"age": NaN}""", """{"age": Infinity}""",
+      """{"age": -Infinity}""", """{"age": +INF}""", """{"age": -INF}""")
+    testCases.foreach { str =>
+      val rdd = spark.sparkContext.parallelize(Seq(str))
+      val df = spark.read.option("allowNonNumericNumbers", "false").json(rdd)
+
+      assert(df.schema.head.name == "_corrupt_record")
+    }
+
+    // quoted non-numeric numbers should still work even allowNonNumericNumbers is off.
+    testCases = Seq("""{"age": "NaN"}""", """{"age": "Infinity"}""", """{"age": "-Infinity"}""")
+    val tests: Seq[Double => Boolean] = Seq(_.isNaN, _.isPosInfinity, _.isNegInfinity)
+
+    testCases.zipWithIndex.foreach { case (str, idx) =>
+      val rdd = spark.sparkContext.parallelize(Seq(str))
+      val df = spark.read.option("allowNonNumericNumbers", "false").json(rdd)
+
+      assert(df.schema.head.name == "age")
+      assert(tests(idx)(df.first().getDouble(0)))
```

Comment:
from @rxin comment that we want to support quoted non-numeric numbers when `allowNonNumericNumbers` is off.